### ISSUE TITLE
Dart CLI commands hang on a slow/unstable internet connection due to analytics

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#56106]
To reproduce, intentionally make your internet suck or be unstable. Try a subway or something :)

Then, any command that sends analytics, i.e. `dart run` or `dart test` hangs for a period of time.

To "fix", disable analytics (which ironically itself hangs for that same period):
```sh
dart --disable-analytics
```

Analytics collection should run "out of band", that is, it should not be blocking (I would not use `await`). I might go an additional step further and say that perhaps the actual mechanism that sends the analytics to a server should happen elsewhere, i.e. I should _never_ have this happen:
```sh
dart run empty_file_with_void_main.dart
# ... script clearly executes, but 'dart' appears to hang for 10-30s
```

**Comment by matanlurey** [comment]
I might check with the `flutter` CLI folks to see how they handle it.

**Comment by devoncarew** [comment]
In our previous implementation of analytics it was possible to get a future for the last item sent, wait on that for a brief time w/ a 100ms-200ms timeout, and exit the process if there was a delay in sending (to not hang the VM exit waiting for analytics). We've since re-written the analytics implementation; I'm not sure who's responsible for the integration into the dart cli, but @bkonyi would be the best person to start with.

**Comment by bkonyi** [comment]
> In our previous implementation of analytics it was possible to get a future for the last item sent, wait on that for a brief time w/ a 100ms-200ms timeout, and exit the process if there was a delay in sending (to not hang the VM exit waiting for analytics). We've since re-written the analytics implementation; I'm not sure who's responsible for the integration into the dart cli, but @bkonyi would be the best person to start with.

Yeah, I'm responsible for the CLI. I thought Elias had done some work to account for this (I know the legacy analytics timeout behavior was brought up), but maybe this never ended up happening.

**Comment by christopherfujino** [comment]
So this call: https://github.com/dart-lang/sdk/blob/main/pkg/dartdev/lib/dartdev.dart#L262

Should default to timing out after 250ms: https://github.com/dart-lang/tools/blob/main/pkgs/unified_analytics/lib/src/analytics.dart#L548

I wonder if there is some other code path by which we don't run this `finally`? Or some other async call to unified_analytics is hanging, before we get to the finally?

**Comment by christopherfujino** [comment]
Hmm, when we timeout, we call `_gaClient.close();`, but I'm thinking maybe that doesn't cancel our pending requests?

**Comment by matanlurey** [comment]
If I trace the code, it eventually gets to [`<IOClient>.close()`](https://pub.dev/documentation/http/latest/io_client/IOClient/close.html), which does a `.close(force: true`), which _should_ cancel pending requests.

**Comment by christopherfujino** [comment]
> If I trace the code, it eventually gets to [`<IOClient>.close()`](https://pub.dev/documentation/http/latest/io_client/IOClient/close.html), which does a `.close(force: true`), which _should_ cancel pending requests.

Ok, that makes sense. Also, assuming the flutter cli tool is not susceptible to this bug, then presumably `Analytics.close()` is closing these connections.

**Comment by zer09** [comment]
It also happen to me on a newly created project, it consistently freeze for 20 seconds before exiting when analytics is enable.

![image](https://github.com/user-attachments/assets/bb2de423-9591-46ee-9245-e933388e46b0)

CC @DanTup 
ref: https://github.com/Dart-Code/Dart-Code/issues/3270

**Comment by mkorbel1** [comment]
Hi, any updates on this?

**Comment by ChaserVasya** [comment]
In my case dart pub hangs are correlated with telemetry network calls:
- the problematic host is www.google-analytics.com
- dart pub get is the most common trigger (also seen with upgrade / outdated)
-  connect to :443 succeeds, but TLS handshake stalls (SSL connection timeout)

From this network path:

- https://pub.dev works
- https://www.google.com works
- https://www.google-analytics.com/mp/collect times out at TLS handshake

So this looks like CLI waiting on telemetry transport when GA endpoint is reachable at TCP level but handshake does not complete. This may help narrow the fix to analytics send path / timeout behavior.

I am from Russia. May be it is DPI affect.
